src/Lykke.Service.PayInternal.Services/BtcTransferRequestService.cs

```diff
@@ -0,0 +1,428 @@
+                        from w in wallets
+                        select new SourceAmount
+                        {
+                            Address = w.Address,
+                            Amount = 0
+                        }, 0, wallets));
+                }
+                else
+                {
+                    sources.AddRange(CalculateSources(transaction.SourceAmounts, 0, wallets));
+                }
+
+                if (!sources.Any())
+                {
+                    await _log.WriteWarningAsync(nameof(BtcTransferRequestService), nameof(CreateTransferAsync),
+                        transferRequest.ToJson(),
+                        "Source addresses don't found or amount is not enought.");
+                    return BtcTransferRequest.CreateErrorTransferRequest(transferRequest.MerchantId,
+                        TransferStatusError.InvalidAmount, transferRequest.TransactionRequests);
+                }
+                var transferWallets = new List<string>(sources.Select(s => s.Address))
+                {
+                    transaction.DestinationAddress
+                };
+
+                var isAddressesValid = CheckAddressesValid(transferWallets);
+                if (!isAddressesValid)
+                {
+                    await _log.WriteWarningAsync(nameof(BtcTransferRequestService), nameof(CreateTransferAsync),
+                        transferRequest.ToJson(),
+                        "Source addresses don't found or amount is not enought.");
+                    return BtcTransferRequest.CreateErrorTransferRequest(transferRequest.MerchantId,
+                        TransferStatusError.InvalidAddress, transferRequest.TransactionRequests);
+                }
+                var result = await CreateBtcTransfer(sources, transaction.DestinationAddress);
+                if (result.Item2 != TransferStatusError.NotError)
+                {
+                    await _log.WriteWarningAsync(nameof(BtcTransferRequestService), nameof(CreateTransferAsync),
+                        transferRequest.ToJson(),
+                        "Create transfer error occurs.");
+                    return BtcTransferRequest.CreateErrorTransferRequest(transferRequest.MerchantId,
+                        result.Item2, transferRequest.TransactionRequests);
+                }
+                transactions.Add(result.Item1);
+            }
+
+            return BtcTransferRequest.CreateTransferRequest(transferRequest.MerchantId, transactions);
+        }
+
+        #endregion
+
+        #region private
+
+        /// <summary>
+        /// Looks through the source addresses and check em up to belong to the Merchant.
+        /// </summary>
+        /// <param name="transferRequest">Transfer request.</param>
+        /// <returns>Error transfer request OR null if validation ended up with success.</returns>
+        private async Task<ITransferRequest> CheckupSourcesAndWalletsAsync(ITransferRequest transferRequest)
```

Comment:
How do you unit-test this code as it is private?
I believe it'd be much better to put it inside separate validation service

Reply:
Currently, no how. But from the other side, unit testing is not fully applicable in this case: we have an external dependency in the method.